openfl/component/aggregator/aggregator.py

```diff
@@ -154,8 +158,21 @@ def __init__(
         # TODO: Remove. Used in deprecated interactive and native APIs
         self.best_tensor_dict: dict = {}
         self.last_tensor_dict: dict = {}
+        # these enable getting all tensors for a task
+        self.collaborator_tasks_results = {}  # {TaskResultKey: list of TensorKeys}
+        self.collaborator_task_weight = {}  # {TaskResultKey: data_size}
+        
 
-        if initial_tensor_dict:
+        # maintain a list of collaborators that have completed task and
+        # reported results in a given round
+        self.collaborators_done = []
+        # Initialize a lock for thread safety
+        self.lock = Lock()
+        self.use_delta_updates = use_delta_updates
+
+        if self._recover():
```

Comment:
Again, this should be determined by the addition of a `experiment_checkpoint` configuration in plan.yaml:

i.e. 
```
aggregator:
...
    experiment_checkpoint: True
```

```suggestion
        if experiment_checkpoint:
            self._recover()
```

Reply:
In addition, there needs to be some kind of check that the previously serialized state matches the current plan. This can be accomplished by storing the `federation_uuid`, which is determined by taking a hash of the plan.yaml file. Changing a plan in the aggregator's current workspace, then restoring state with the older configuration could result in unexpected behavior.

Reply:
is it a valid to change plan on a running experiment?
what shall be done in case of a mismatch? drop data and start from scratch?  does it require notifying the collaborators?

Reply:
@psfoley can we discuss it and implement it in a separate PR?

Reply:
> is it valid to change plan on a running experiment? what shall be done in case of a mismatch? drop data and start from scratch? does it require notifying the collaborators?

I think the most straightforward and safest response would be to raise an error at the aggregator in the event of a mismatch to block the restart. If the experiment is allowed to restart with a plan mismatch, it would eventually be terminated when the collaborator tries to verify the response ([see](https://github.com/securefederatedai/openfl/blob/develop/openfl/transport/grpc/aggregator_client.py#L347). Adding an additional check before communication even allowed is a good thing

Reply:
After internal conversation it was agreed that looking for changes in the plan before reloading state was needed but can be pushed to another PR. Resolving this comment for now.